File: algorithms/algorithms-cplusplus/leetcode/Question_0889_Construct_Binary_Tree_from_Preorder_and_Postorder_Traversal.cc
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  TreeNode* constructFromPrePost(vector<int>& preorder, vector<int>& postorder) {
    int preIndex = 0, posIndex = 0;
    return help(preorder, preIndex, postorder, posIndex);
  }

 private:
  TreeNode* help(vector<int>& preorder, int& preIndex, vector<int>& postorder, int& posIndex) {
    TreeNode* root = new TreeNode(preorder[preIndex++]);
    if (root->val != postorder[posIndex]) {
      root->left = help(preorder, preIndex, postorder, posIndex);
    }
    if (root->val != postorder[posIndex]) {
      root->right = help(preorder, preIndex, postorder, posIndex);
    }
    posIndex++;
    return root;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0889_Construct_Binary_Tree_from_Preorder_and_Postorder_Traversal.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0889_Construct_Binary_Tree_from_Preorder_and_Postorder_Traversal {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private int preIndex = 0, posIndex = 0;

    public TreeNode constructFromPrePost(int[] preorder, int[] postorder) {
        TreeNode root = new TreeNode(preorder[preIndex++]);
        if (root.val != postorder[posIndex]) {
            root.left = constructFromPrePost(preorder, postorder);
        }
        if (root.val != postorder[posIndex]) {
            root.right = constructFromPrePost(preorder, postorder);
        }
        posIndex++;
        return root;
    }
}
